From 89d57995c8a5a08dfc82b7636d7a6d688c0645ac Mon Sep 17 00:00:00 2001
From: Ganesh prasad Sahu <GaneshPrasad_Sahu@comcast.com>
From: E. Ocaña González <eocanha@igalia.com>
Date: Thu, 25 Jun 2026 05:44:42 -0700
Subject: [PATCH] [libWebRTC][GStreamer] Missing video in WebRTC playback when
 decoding on playback pipeline https://bugs.webkit.org/show_bug.cgi?id=317111

Reviewed by Xabier Rodriguez-Calvar.

LibWebRTC has a dedicated Decoder component that is supposed to decode video
frames and supply them like that to its observers. Under those assumptions,
missing some of those decoded frames (eg: because no observer is listening yet)
isn't very problematic in this expected case.

However, in order to facilitate hardware decoding, the libWebRTC Decoder
implementation provided by the GStreamer port emits encoded frames (that is,
doesn't decode at all) unders some circumstances, with the hope that those
encoded frames end up reaching the playback pipeline and being decoded there.

In such a case, missing video frames because no observer has been registered
(a circumstance that has been reported by at least one downstream setup but
that can't be reproduced reliably) can be problematic. Missing the first
"frames" (buffers) of an encoded video means that the video headers (eg:
SPS/PPS for H.264) are missed that's a condition the actual video decoder
can't recover from.

This patch provides a mechanism to keep undelivered buffers when they come
in an encoded format and there's no registered observer to deliver them to,
and redeliver them when the first observer is registered. This solves the
problem detected in the mentioned downstream setup.

If a format change is detected or if RealtimeMediaSource is stopping, the
list of pending frames is emtied, since there's no point anymore on
delivering them.

Co-authored by: Ganesh prasad Sahu <GaneshPrasad_Sahu@comcast.com>

* Source/WebCore/platform/VideoFrame.h:
(WebCore::VideoFrame::isEncoded const): Added generic implementation to know if a VideoFrame has an encoded format or not. Defaults to false.
(WebCore::VideoFrame::hasSameEncodedFormat const): Compares if format of one sample is the same as the one from another. Defaults to false.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
(WebCore::VideoFrameGStreamer::VideoFrameGStreamer): Added assertion to ensure m_sample is not null.
(WebCore::VideoFrameGStreamer::isEncoded const): Specific implementation, only true when the caps are other than video/x-raw.
(WebCore::VideoFrameGStreamer::hasSameEncodedFormat const): Format comparison is implemented by comparing caps.
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h: Added isEncoded() declaration.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::videoFrameAvailable): Added code to hoard decoded frames when there's no observer, and redeliver then later. The frame delivery code is reused from the regular case, so it has been factored into a lambda. If a format change is detected, the pending frames are emptied, because there's no point on redelivering them (only the ones with the new format are relevant).
(WebCore::RealtimeMediaSource::stop): Empty the pending frames, since there's no point in delivering them when we're stopping the whole RealtimeMediaSource.
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h: Added PendingVideoFrame struct to hold the pending frames and metadata, a Vector of pending frames, and the constant that defines the maximum number of frames that can be kept waiting redelivery.

Canonical link: https://commits.webkit.org/315815@main
---
 Source/WebCore/platform/VideoFrame.h          |  2 +
 .../gstreamer/VideoFrameGStreamer.cpp         | 33 ++++++++++++++++
 .../graphics/gstreamer/VideoFrameGStreamer.h  |  3 ++
 .../mediastream/RealtimeMediaSource.cpp       | 39 ++++++++++++++++++-
 .../mediastream/RealtimeMediaSource.h         |  7 ++++
 5 files changed, 82 insertions(+), 2 deletions(-)

diff --git a/Source/WebCore/platform/VideoFrame.h b/Source/WebCore/platform/VideoFrame.h
index 432626152354b..b95eb295cda90 100644
--- a/Source/WebCore/platform/VideoFrame.h
+++ b/Source/WebCore/platform/VideoFrame.h
@@ -64,6 +64,8 @@ class VideoFrame : public ThreadSafeRefCounted<VideoFrame> {
 
     virtual FloatSize presentationSize() const = 0;
     virtual uint32_t pixelFormat() const = 0;
+    virtual bool isEncoded() const { return false; }
+    virtual bool hasSameEncodedFormat(const VideoFrame&) const { return false; }
 
     virtual bool isRemoteProxy() const { return false; }
     virtual bool isLibWebRTC() const { return false; }
diff --git a/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp
index 3a9555ec7d45c..9bb7f2710bdad 100644
--- a/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp
@@ -120,6 +120,7 @@ VideoFrameGStreamer::VideoFrameGStreamer(GRefPtr<GstSample>&& sample, const Floa
     , m_presentationSize(presentationSize)
 {
     ASSERT(m_sample);
+
     GstBuffer* buffer = gst_sample_get_buffer(m_sample.get());
     RELEASE_ASSERT(buffer);
 
@@ -220,6 +221,38 @@ RefPtr<JSC::Uint8ClampedArray> VideoFrameGStreamer::computeRGBAImageData() const
     return JSC::Uint8ClampedArray::tryCreate(WTFMove(bufferStorage), 0, byteLength);
 }
 
+bool VideoFrameGStreamer::isEncoded() const
+{
+    GstCaps* caps = gst_sample_get_caps(m_sample.get());
+    if (!caps)
+        return false;
+
+    const GstStructure* structure = gst_caps_get_structure(caps, 0);
+    if (!structure)
+        return false;
+
+    return !g_strcmp0(gst_structure_get_name(structure), "video/x-raw");
+}
+
+bool VideoFrameGStreamer::hasSameEncodedFormat(const VideoFrame& other) const
+{
+    if (!other.isGStreamer())
+        return false;
+
+    GstCaps* thisCaps = gst_sample_get_caps(m_sample.get());
+    GstCaps* otherCaps = gst_sample_get_caps(static_cast<const VideoFrameGStreamer&>(other).m_sample.get());
+
+    if (!thisCaps && !otherCaps)
+        return true;
+
+    if (!thisCaps || !otherCaps)
+        return false;
+
+    ASSERT(thisCaps && otherCaps);
+
+    return gst_caps_is_equal(thisCaps, otherCaps);
+}
+
 } // namespace WebCore
 
 #endif // ENABLE(VIDEO) && USE(GSTREAMER)
diff --git a/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h b/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h
index 2a641435c78ff..bf6a9f37e4fcf 100644
--- a/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h
+++ b/Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h
@@ -60,6 +60,9 @@ class VideoFrameGStreamer final : public VideoFrame {
     GstSample* sample() const { return m_sample.get(); }
     RefPtr<JSC::Uint8ClampedArray> computeRGBAImageData() const;
 
+    bool isEncoded() const final;
+    bool hasSameEncodedFormat(const VideoFrame&) const final;
+
 private:
     VideoFrameGStreamer(GRefPtr<GstSample>&&, const FloatSize& presentationSize, const MediaTime& presentationTime = MediaTime::invalidTime(), Rotation = Rotation::None, bool videoMirrored = false, std::optional<VideoFrameTimeMetadata>&& = std::nullopt);
     VideoFrameGStreamer(const GRefPtr<GstSample>&, const MediaTime& presentationTime, Rotation = Rotation::None);
diff --git a/Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp b/Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp
index 5ebc2b98ca639..26038a5c6b258 100644
--- a/Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp
+++ b/Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp
@@ -218,8 +218,37 @@ void RealtimeMediaSource::videoFrameAvailable(VideoFrame& videoFrame, VideoFrame
     updateHasStartedProducingData();
 
     Locker locker { m_videoFrameObserversLock };
-    for (auto* observer : m_videoFrameObservers)
-        observer->videoFrameAvailable(videoFrame, metadata);
+
+    // videoFrameObservers needs to be passed as parameter because if m_videoFrameObservers
+    // is used, that confuses the thread safety static analysis, which can't deduce that
+    // the lock has been taken by the outer scope of the videoFrameAvailable() method.
+    auto deliverFrame = [&] (Ref<VideoFrame>&& frame, VideoFrameTimeMetadata&& frameMetadata,
+        HashSet<VideoFrameObserver*>& videoFrameObservers) {
+        for (auto* observer : videoFrameObservers)
+            observer->videoFrameAvailable(frame, frameMetadata);
+    };
+
+    // Store encoded (H.264, vp9, etc.) video frames to avoid missing the headers when the first observer starts listening.
+    if (m_videoFrameObservers.isEmpty() && videoFrame.isEncoded()) {
+        // If the encoded format has changed, empty the pending frames, because there's no point on delivering them anymore.
+        if (!m_pendingVideoFrames.isEmpty() && !(Ref(m_pendingVideoFrames.last().frame)->hasSameEncodedFormat(videoFrame)))
+            m_pendingVideoFrames.clear();
+
+        if (m_pendingVideoFrames.size() < maxPendingVideoFramesBeforeAddTrack)
+            m_pendingVideoFrames.append(PendingVideoFrame { Ref(videoFrame), WTFMove(metadata) });
+        else
+            ALWAYS_LOG_IF(m_logger, LOGIDENTIFIER, "RealtimeMediaSource: Dropping video frame (queue is full) ", m_pendingVideoFrames.size(), " frames");
+        return;
+    }
+
+    if (!m_pendingVideoFrames.isEmpty()) {
+        ALWAYS_LOG_IF(m_logger, LOGIDENTIFIER, "RealtimeMediaSource: Delivering ", m_pendingVideoFrames.size(), " queued frame(s) (observer ready)");
+        for (auto& pending : m_pendingVideoFrames)
+            deliverFrame(WTFMove(pending.frame), WTFMove(pending.metadata), m_videoFrameObservers);
+        m_pendingVideoFrames.clear();
+    }
+
+    deliverFrame(Ref(videoFrame), WTFMove(metadata), m_videoFrameObservers);
 }
 
 void RealtimeMediaSource::audioSamplesAvailable(const MediaTime& time, const PlatformAudioData& audioData, const AudioStreamDescription& description, size_t numberOfFrames)
@@ -257,6 +286,12 @@ void RealtimeMediaSource::stop()
     ALWAYS_LOG_IF(m_logger, LOGIDENTIFIER);
 
     m_isProducingData = false;
+
+    {
+        Locker locker { m_videoFrameObserversLock };
+        m_pendingVideoFrames.clear();
+    }
+
     stopProducingData();
 }
 
diff --git a/Source/WebCore/platform/mediastream/RealtimeMediaSource.h b/Source/WebCore/platform/mediastream/RealtimeMediaSource.h
index 091aa0ead2afe..c7ee585d9fd11 100644
--- a/Source/WebCore/platform/mediastream/RealtimeMediaSource.h
+++ b/Source/WebCore/platform/mediastream/RealtimeMediaSource.h
@@ -300,6 +300,13 @@ class WEBCORE_EXPORT RealtimeMediaSource
     mutable Lock m_videoFrameObserversLock;
     HashSet<VideoFrameObserver*> m_videoFrameObservers WTF_GUARDED_BY_LOCK(m_videoFrameObserversLock);
 
+    struct PendingVideoFrame {
+        Ref<VideoFrame> frame;
+        VideoFrameTimeMetadata metadata;
+    };
+    static constexpr size_t maxPendingVideoFramesBeforeAddTrack = 30;
+    Vector<PendingVideoFrame> m_pendingVideoFrames WTF_GUARDED_BY_LOCK(m_videoFrameObserversLock);
+
     // Set on the main thread from constraints.
     IntSize m_size;
     // Set on sample generation thread.
